include/rmm/mr/device/tracking_resource_adaptor.hpp

```diff
@@ -0,0 +1,288 @@
+   */
+  bool supports_streams() const noexcept override { return upstream_->supports_streams(); }
+
+  /**
+   * @brief Query whether the resource supports the get_mem_info API.
+   *
+   * @return bool true if the upstream resource supports get_mem_info, false otherwise.
+   */
+  bool supports_get_mem_info() const noexcept override
+  {
+    return upstream_->supports_get_mem_info();
+  }
+
+  /**
+   * @brief Get the outstanding allocations map
+   *
+   * @return std::map<void*, allocation_info> const& of a map of allocations. The key
+   * is the allocated memory pointer and the data is the allocation_info structure, which
+   * contains size and, potentially, stack traces.
+   */
+  std::map<void*, allocation_info> const& get_outstanding_allocations() const
+  {
+    return allocations;
+  }
+
+  /**
+   * @brief Query the number of bytes that have been allocated. Note that
+   * this can not be used to know how large of an allocation is possible due
+   * to both possible fragmentation and also internal page sizes and alignment
+   * that is not tracked by this allocator.
+   *
+   * @return std::size_t number of bytes that have been allocated through this
+   * allocator.
+   */
+  std::size_t get_allocated_bytes() const { return allocated_bytes_; }
+
+  /**
+   * @brief Log any outstanding allocations via RMM_LOG_DEBUG
+   *
+   */
+  void log_outstanding_allocations() const
+  {
+#if SPDLOG_ACTIVE_LEVEL <= SPDLOG_LEVEL_DEBUG
+    read_lock_t lock(mtx);
+    if (not allocations.empty()) {
+      std::ostringstream oss;
+      for (auto const& al : allocations) {
+        oss << al.first << ": " << al.second.allocation_size << " B";
+        if (al.second.strace != nullptr) {
+          oss << " : callstack:" << std::endl << *al.second.strace;
+        }
+        oss << std::endl;
+      }
+      RMM_LOG_DEBUG("Outstanding Allocations: {}", oss.str());
+    }
+#endif  // SPDLOG_ACTIVE_LEVEL <= SPDLOG_LEVEL_DEBUG
+  }
+
+  /**
+   * @brief Get the number of outstanding allocations
+   *
+   * @return std::size_t number of allocations still outstanding
+   */
+  std::size_t get_num_outstanding_allocations() const { return allocations.size(); };
+
+ private:
+  /**
+   * @brief Allocates memory of size at least `bytes` using the upstream
+   * resource as long as it fits inside the allocation limit.
+   *
+   * The returned pointer has at least 256B alignment.
+   *
+   * @throws `rmm::bad_alloc` if the requested allocation could not be fulfilled
+   * by the upstream resource.
+   *
+   * @param bytes The size, in bytes, of the allocation
+   * @param stream Stream on which to perform the allocation
+   * @return void* Pointer to the newly allocated memory
+   */
+  void* do_allocate(std::size_t bytes, cudaStream_t stream) override
+  {
+    void* p = [&]() {
+      try {
+        return upstream_->allocate(bytes, stream);
+      } catch (std::exception const& e) {
+        RMM_LOG_ERROR("[A][Stream {}][Upstream {}B][FAILURE {}]",
```

Comment:
Undecided on whether this is needed. We haven't added logging to all MRs yet, so this is helpful now. But ultimately if an MR is throwing an error, it should be logging that error itself. This would cause duplicate log messages with, for example, pool MRs.

Reply:
I think we discussed removing it in the past. I'm not sure where that went, but I will take it out.